subxt/src/events.rs

```diff
@@ -72,108 +414,66 @@ impl RawEvent {
-                        variant_index,
-                        data: event_data.into(),
-                    };
-
-                    // topics come after the event data in EventRecord
-                    let topics = Vec::<T::Hash>::decode(input)?;
-                    log::debug!("topics: {:?}", topics);
-
-                    event
-                }
-                Err(err) => return Err(err),
-            };
-            r.push((phase.clone(), raw));
-        }
-        Ok(r)
-    }
-
-    fn decode_raw_event(
-        &self,
-        event_metadata: &EventMetadata,
-        input: &mut &[u8],
-        output: &mut Vec<u8>,
-    ) -> Result<(), BasicError> {
-        log::debug!(
-            "Decoding Event '{}::{}'",
-            event_metadata.pallet(),
-            event_metadata.event()
-        );
-        for arg in event_metadata.variant().fields() {
-            let type_id = arg.ty().id();
-            self.decode_type(type_id, input, output)?
-        }
-        Ok(())
-    }
-
-    fn decode_type(
-        &self,
-        type_id: u32,
-        input: &mut &[u8],
-        output: &mut Vec<u8>,
-    ) -> Result<(), BasicError> {
+// Attempt to dynamically decode a single event from our events input.
+fn decode_raw_event_details<T: Config>(
+    metadata: &Metadata,
+    index: u32,
+    input: &mut &[u8],
+) -> Result<RawEventDetails, BasicError> {
+    // Decode basic event details:
+    let phase = Phase::decode(input)?;
+    let pallet_index = input.read_byte()?;
+    let variant_index = input.read_byte()?;
+    log::debug!(
+        "phase {:?}, pallet_index {}, event_variant: {}",
+        phase,
+        pallet_index,
+        variant_index
+    );
+    log::debug!("remaining input: {}", hex::encode(&input));
+
+    // Get metadata for the event:
+    let event_metadata = metadata.event(pallet_index, variant_index)?;
+    log::debug!(
+        "Decoding Event '{}::{}'",
+        event_metadata.pallet(),
+        event_metadata.event()
+    );
+
+    // Use metadata to figure out which bytes belong to this event:
+    let mut event_bytes = Vec::new();
+    for arg in event_metadata.variant().fields() {
+        let type_id = arg.ty().id();
         let all_bytes = *input;
         // consume some bytes, moving the cursor forward:
-        decode_and_consume_type(type_id, &self.metadata.runtime_metadata().types, input)?;
+        decode_and_consume_type(type_id, &metadata.runtime_metadata().types, input)?;
         // count how many bytes were consumed based on remaining length:
         let consumed_len = all_bytes.len() - input.len();
         // move those consumed bytes to the output vec unaltered:
-        output.extend(&all_bytes[0..consumed_len]);
-        Ok(())
+        event_bytes.extend(&all_bytes[0..consumed_len]);
     }
+
+    // topics come after the event data in EventRecord. They aren't used for
+    // anything at the moment, so just decode and throw them away.
```

Comment:
Reading https://github.com/paritytech/substrate/pull/2491 I wonder if `topics` will become a topic again (sorry) now that smoldot is a thing?

Reply:
At least we properly decode them, so things won't break if they do start being used! (though I expect other stuff may have changed that would be breaking by then...)